Printing result of a void function parses, but doesn't compile in C++

#### Summary:
Evaluating a void function within a print statement parses, but does not compile.

#### Reproducible Steps:
```stan
functions {
  void foo() { }
}
transformed data {
  print(foo());
}
parameters {
  real<lower = 0, upper = 1> theta;
}
model { }
```

#### Current Output:
C++ compiler error:
```
In file included from <built-in>:336:
In file included from <command line>:7:
./foo.hpp:90:17: error: no matching function for call to 'stan_print'
                stan_print(pstream__,foo(pstream__));
                ^~~~~~~~~~
stan/lib/stan_math/stan/math/rev/mat/fun/stan_print.hpp:10:17: note: candidate function not viable:
      cannot convert argument of incomplete type 'void' to 'const stan::math::var'
    inline void stan_print(std::ostream* o, const var& x) {
                ^
stan/lib/stan_math/stan/math/prim/mat/fun/stan_print.hpp:12:10: note: candidate template ignored:
      substitution failure [with T = void]: cannot form a reference to 'void'
    void stan_print(std::ostream* o, const T& x) {
         ^                                  ~
stan/lib/stan_math/stan/math/prim/mat/fun/stan_print.hpp:17:10: note: candidate template ignored:
      could not match 'vector<type-parameter-0-0, allocator<type-parameter-0-0> >' against 'void'
    void stan_print(std::ostream* o, const std::vector<T>& x) {
         ^
stan/lib/stan_math/stan/math/prim/mat/fun/stan_print.hpp:27:10: note: candidate template ignored:
      could not match 'Matrix<type-parameter-0-0, -1, 1, 0, -1, 1>' against 'void'
    void stan_print(std::ostream* o,
         ^
stan/lib/stan_math/stan/math/prim/mat/fun/stan_print.hpp:38:10: note: candidate template ignored:
      could not match 'Matrix<type-parameter-0-0, 1, -1, 1, 1, -1>' against 'void'
    void stan_print(std::ostream* o,
         ^
stan/lib/stan_math/stan/math/prim/mat/fun/stan_print.hpp:49:10: note: candidate template ignored:
      could not match 'Matrix<type-parameter-0-0, -1, -1, 0, -1, -1>' against 'void'
    void stan_print(std::ostream* o,
         ^
1 error generated.
```

#### Expected Output:
Either:
- disallow printing of `void`
or
- allow it and print `void` or something similar

#### Current Version:
v2.13.1

--- a/src/docs/stan-reference/appendices.tex
+++ b/src/docs/stan-reference/appendices.tex
@@ -775,6 +775,10 @@ Both the conditional if-then-else statement and while-loop statement
 require the expression denoting the condition to be a primitive type,
 integer or real.
 
+\subsection{Print Arguments}
+
+The arguments to a print statement cannot be void.
+
 \subsection{Only Break and Continue in Loops}
 
 The \code{break} and \code{continue} statements may only be used

--- a/src/docs/stan-reference/language.tex
+++ b/src/docs/stan-reference/language.tex
@@ -3509,7 +3509,11 @@ u[1] = [1, 4]
 u[2] = [9, 16]
 \end{stancode}
 
+\subsection{Non-void Input}
 
+The input type to a print function cannot be void.  In particular, it
+can't be the result of a user-defined void function.  All other types
+are allowed as arguments to the print function.
 
 \subsection{Print Frequency}
 

--- a/src/stan/lang/grammars/semantic_actions.hpp
+++ b/src/stan/lang/grammars/semantic_actions.hpp
@@ -807,6 +807,11 @@ namespace stan {
     };
     extern boost::phoenix::function<validate_in_loop> validate_in_loop_f;
 
+    struct non_void_expression : public phoenix_functor_ternary {
+      void operator()(const expression& e, bool& pass,
+                      std::ostream& error_msgs) const;
+    };
+    extern boost::phoenix::function<non_void_expression> non_void_expression_f;
   }
 }
 #endif

--- a/src/stan/lang/grammars/semantic_actions_def.cpp
+++ b/src/stan/lang/grammars/semantic_actions_def.cpp
@@ -2619,6 +2619,17 @@ namespace stan {
     }
     boost::phoenix::function<validate_in_loop> validate_in_loop_f;
 
+    void non_void_expression::operator()(const expression& e, bool& pass,
+                                         std::ostream& error_msgs) const {
+      // ill-formed shouldn't be possible, but just in case
+      pass = e.expression_type().type() != VOID_T
+        && e.expression_type().type() != ILL_FORMED_T;
+      if (!pass)
+        error_msgs << "ERROR:  expected printable (non-void) expression."
+                   << std::endl;
+    }
+    boost::phoenix::function<non_void_expression> non_void_expression_f;
+
   }
 }
 

--- a/src/stan/lang/grammars/statement_grammar_def.hpp
+++ b/src/stan/lang/grammars/statement_grammar_def.hpp
@@ -217,7 +217,8 @@ namespace stan {
       printable_r.name("printable");
       printable_r
         %= printable_string_r
-        | expression_g(_r1);
+        | expression_g(_r1)
+          [non_void_expression_f(_1, _pass, boost::phoenix::ref(error_msgs_))];
 
       printable_string_r.name("printable quoted string");
       printable_string_r

--- a/src/test/test-models/bad/print-void.stan
+++ b/src/test/test-models/bad/print-void.stan
@@ -1,0 +1,10 @@
+functions {
+  void foo(int n) {
+  }
+}
+transformed data {
+  int z;
+  z = 2;
+  print(foo(z));
+}
+model { }

--- a/src/test/unit/lang/parser/print_function_test.cpp
+++ b/src/test/unit/lang/parser/print_function_test.cpp
@@ -1,0 +1,7 @@
+#include <gtest/gtest.h>
+#include <test/unit/lang/utility.hpp>
+
+TEST(langParser, printVoidBad) {
+  test_throws("print-void", 
+              "ERROR:  expected printable (non-void) expression.");
+}

